src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsassessforearlyreleaseapi/entity/AssessmentStatus.kt

```diff
@@ -285,16 +138,436 @@ enum class AssessmentStatus {
+    }
+    on<AssessmentLifecycleEvent.SubmitForAddressChecks> {
+      transitionTo(AssessmentState.AwaitingAddressAndRiskChecks)
+    }
+    on<AssessmentLifecycleEvent.StartAddressChecks> {
+      transitionTo(AssessmentState.AddressAndRiskChecksInProgress)
+    }
+    on<AssessmentLifecycleEvent.SubmitForDecision> {
+      transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+    }
+    on<AssessmentLifecycleEvent.Approve> {
+      transitionTo(AssessmentState.Approved)
+    }
+    on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+      transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+      transitionTo(AssessmentState.PassedPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.FailAddressChecks> {
+      transitionTo(AssessmentState.AddressUnsuitable)
+    }
+    on<AssessmentLifecycleEvent.Refuse> {
+      transitionTo(AssessmentState.Refused)
+    }
+    on<AssessmentLifecycleEvent.OptOut> {
+      transitionTo(AssessmentState.OptedOut(this))
+    }
+  }
+
+  state<AssessmentState.Postponed> {
+    on<AssessmentLifecycleEvent.SubmitForDecision> {
+      transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+    }
+    on<AssessmentLifecycleEvent.Approve> {
+      transitionTo(AssessmentState.Approved)
+    }
+    on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+      transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+      transitionTo(AssessmentState.PassedPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.Timeout> {
+      transitionTo(AssessmentState.TimedOut)
+    }
+  }
+
+  state<AssessmentState.OptedOut> {
+    on<AssessmentLifecycleEvent.CompleteEligibilityAndSuitability> {
+      transitionTo(AssessmentState.EligibleAndSuitable)
+    }
+    on<AssessmentLifecycleEvent.SubmitForAddressChecks> {
+      transitionTo(AssessmentState.AwaitingAddressAndRiskChecks)
+    }
+    on<AssessmentLifecycleEvent.StartAddressChecks> {
+      transitionTo(AssessmentState.AddressAndRiskChecksInProgress)
+    }
+    on<AssessmentLifecycleEvent.SubmitForDecision> {
+      transitionTo(AssessmentState.AwaitingPreDecisionChecks)
+    }
+    on<AssessmentLifecycleEvent.Approve> {
+      transitionTo(AssessmentState.Approved)
+    }
+    on<AssessmentLifecycleEvent.CompleteAddressChecks> {
+      transitionTo(AssessmentState.AwaitingPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.ReleaseOnHDC> {
+      transitionTo(AssessmentState.PassedPreReleaseChecks)
+    }
+    on<AssessmentLifecycleEvent.FailAddressChecks> {
+      transitionTo(AssessmentState.AddressUnsuitable)
+    }
+    on<AssessmentLifecycleEvent.Refuse> {
+      transitionTo(AssessmentState.Refused)
+    }
+    on<AssessmentLifecycleEvent.Timeout> {
+      transitionTo(AssessmentState.TimedOut)
+    }
+    on<AssessmentLifecycleEvent.OptBackIn> {
+      transitionTo(this.previousState)
```

Comment:
so when opting out, return to the previous state that's associated with the `OptOutState`

at this level - `this` refers to the surrounding state, `it` refers to attributes of the event